zipkin-storage/elasticsearch/src/main/java/zipkin2/elasticsearch/ElasticsearchStorage.java

```diff
@@ -341,41 +357,67 @@ IndexTemplates ensureIndexTemplates() {
+      ensureIndexTemplate(http, buildUrl(templates, AUTOCOMPLETE), templates.autocomplete());
       return templates;
     } catch (IOException e) {
       throw Platform.get().uncheckedIOException(e);
     }
   }
 
-  HttpUrl buildUrl(HttpCall.Factory http, IndexTemplates templates, String type) {
-    HttpUrl.Builder builder = http.baseUrl.newBuilder("_template");
+  String buildUrl(IndexTemplates templates, String type) {
     String indexPrefix = indexNameFormatter().index() + templates.indexTypeDelimiter();
-    return builder.addPathSegment(indexPrefix + type + "_template").build();
+    return "/_template/" + indexPrefix + type + "_template";
   }
 
-  @Memoized // hosts resolution might imply a network call, and we might make a new okhttp instance
-  public HttpCall.Factory http() {
+  @Memoized // a new client factory means new connections
+  ClientFactory clientFactory() {
+    ClientFactoryBuilder builder = new ClientFactoryBuilder()
+      .useHttp2Preface(false);
+    clientFactoryCustomizer().accept(builder);
+    return builder.build();
+  }
+
+  @Memoized // hosts resolution might imply a network call, and we might make a new client instance
+  public HttpClient httpClient() {
     List<String> hosts = hostsSupplier().get();
     if (hosts.isEmpty()) throw new IllegalArgumentException("no hosts configured");
-    OkHttpClient ok =
-        hosts.size() == 1
-            ? client()
-            : client()
-                .newBuilder()
-                .dns(PseudoAddressRecordSet.create(hosts, client().dns()))
-                .build();
-    ok.dispatcher().setMaxRequests(maxRequests());
-    ok.dispatcher().setMaxRequestsPerHost(maxRequests());
-    return new HttpCall.Factory(ok, HttpUrl.parse(hosts.get(0)));
+
+    final String clientUrl;
+    if (hosts.size() == 1) {
+      clientUrl = hosts.get(0);
+    } else {
+      List<URL> urls = hosts.stream()
+        .map(host -> {
+          try {
+            return new URL(host);
+          } catch (MalformedURLException e) {
+            throw new IllegalArgumentException("Invalid host: " + host, e);
+          }
+        })
+        .collect(Collectors.toList());
+      List<Endpoint> endpoints = urls.stream()
+        .map(url -> Endpoint.parse(url.getAuthority()))
+        .collect(Collectors.toList());
+      EndpointGroup group = HttpHealthCheckedEndpointGroup.of(
+        new StaticEndpointGroup(endpoints), "/_cluster/health");
+      EndpointGroupRegistry.register("elasticsearch", group, EndpointSelectionStrategy.ROUND_ROBIN);
+      clientUrl = urls.get(0).getProtocol() + "://group:elasticsearch" + urls.get(0).getPath();
+    }
+
+    HttpClientBuilder client = new HttpClientBuilder(clientUrl)
+      .factory(clientFactory())
+      .decorator(HttpDecodingClient.newDecorator());
+
+    clientCustomizer().accept(client);
+
+    return client.build();
   }
 
-  @Override
-  public void close() {
-    if (!shutdownClientOnClose()) return;
```

Comment:
intentional to remove the shutdownClientOnClose?

Reply:
Yeah without customization, we use the default `ClientFactory`, and it's a no-op to close it so no need for us to worry.